src/pudl/io_managers.py

```diff
@@ -341,6 +341,69 @@ def handle_output(self, context: OutputContext, obj: pd.DataFrame | str):
                 "queries."
             )
 
+    def load_input(self, context: InputContext) -> pd.DataFrame:
+        """Load a dataframe from a sqlite database.
+
+        Args:
+            context: dagster keyword that provides access output information like asset
+                name.
+        """
+        table_name = self._get_table_name(context)
+        _ = self._get_sqlalchemy_table(table_name)
+
+        engine = self.engine
+
+        with engine.connect() as con:
+            try:
+                df = pd.read_sql_table(table_name, con)
+            except ValueError:
+                raise ValueError(
+                    f"{table_name} not found. Either the table was dropped "
+                    "or it doesn't exist in the pudl.metadata.resources."
+                    "Add the table to the metadata and recreate the database."
+                )
+            if df.empty:
+                raise AssertionError(
+                    f"The {table_name} table is empty. Materialize "
+                    "the {table_name} asset so it is available in the database."
+                )
+            return df
+
+
+class PudlSQLiteIOManager(SQLiteIOManager):
+    """IO Manager that writes and retrieves dataframes from a SQLite database.
+
+    This class extends the SQLiteIOManager class to manage database metadata and dtypes
+    using the :func:`pudl.metadata.classes.Package` class.
+    """
+
+    def __init__(
+        self,
+        base_dir: str = None,
+        db_name: str = None,
+        package: Package = None,
+        timeout: float = 1_000.0,
+    ):
+        """Initialize PudlSQLiteIOManager.
+
+        Args:
+            base_dir: base directory where all the step outputs which use this object
+                manager will be stored in.
+            db_name: the name of sqlite database.
+            package: Tabular datapackage used to create sqlalchemy metadata and check
```

Comment:
It's not really a tabular datapackage -- our `Package` class echoes a lot of the parts of that standard but it doesn't actually implement it or allow us to output them / generate datapackage objects.